google/resource_google_folder.go

```diff
@@ -0,0 +1,160 @@
+
+func resourceGoogleFolder() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGoogleFolderCreate,
+		Read:   resourceGoogleFolderRead,
+		Update: resourceGoogleFolderUpdate,
+		Delete: resourceGoogleFolderDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: resourceGoogleFolderImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			// Format is 'folders/{folder_id}.
+			// The folder id holds the same value.
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			// Format is either folders/{folder_id} or organizations/{org_id}.
+			"parent": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			// Must be unique amongst its siblings.
+			"display_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"lifecycle_state": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"create_time": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGoogleFolderCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	displayName := d.Get("display_name").(string)
+	parent := d.Get("parent").(string)
+
+	op, err := config.clientResourceManagerV2Beta1.Folders.Create(&resourceManagerV2Beta1.Folder{
+		DisplayName: displayName,
+	}).Parent(parent).Do()
+
+	if err != nil {
+		return fmt.Errorf("Error creating folder '%s' in '%s': %s", displayName, parent, err)
+	}
+
+	err = resourceManagerV2Beta1OperationWait(config, op, "creating folder")
+
+	if err != nil {
+		return fmt.Errorf("Error creating folder '%s' in '%s': %s", displayName, parent, err)
+	}
+
+	// Since we waited above, the operation is guaranteed to have been successful by this point.
+	waitOp, err := config.clientResourceManager.Operations.Get(op.Name).Do()
+	if err != nil {
+		return fmt.Errorf("The folder '%s' has been created but we could not retrieve its id. Delete the folder manually and retry or use 'terraform import'.", displayName)
+	}
+	response := waitOp.Response.(map[string]interface{})
+	d.SetId(response["name"].(string))
+
+	return resourceGoogleFolderRead(d, meta)
+}
+
+func resourceGoogleFolderRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	folder, err := config.clientResourceManagerV2Beta1.Folders.Get(d.Id()).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, d.Id())
+	}
+
+	d.Set("name", folder.Name)
+	d.Set("parent", folder.Parent)
+	d.Set("display_name", folder.DisplayName)
+	d.Set("lifecycle_state", folder.LifecycleState)
+	d.Set("create_time", folder.CreateTime)
+
+	return nil
+}
+
+func resourceGoogleFolderUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	displayName := d.Get("display_name").(string)
+
+	d.Partial(true)
+	if d.HasChange("display_name") {
+		_, err := config.clientResourceManagerV2Beta1.Folders.Patch(d.Id(), &resourceManagerV2Beta1.Folder{
```

Comment:
Any way you could combine the patch code for display_name and parent into one api call? This would result in less code and be less likely to fail.

Reply:
No we can't.  The display_name must be updated with `Patch` and the parent must be updated with `Move`

Reply:
I missed that, thanks.